pkg/apphealth/health_test.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apphealth
+
+import (
+	"math"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+)
+
+func TestAppHealth_setResult(t *testing.T) {
+	var threshold int32 = 3
+	h := NewAppHealth(&Config{
+		Threshold: threshold,
+	}, nil)
+
+	statusChange := make(chan uint8, 1)
+	unexpectedStatusChanges := atomic.NewUint32(0)
+	h.OnHealthChange(func(status uint8) {
+		select {
+		case statusChange <- status:
+			// Do nothing
+		default:
+			// If the channel is full, it means we were not expecting a status change
+			unexpectedStatusChanges.Inc()
+		}
+	})
+
+	simulateFailures := func(n int32) {
+		statusChange <- 255 // Fill the channel
+		for i := int32(0); i < n; i++ {
+			if i == threshold-1 {
+				<-statusChange // Allow the channel to be written into
+			}
+			h.setResult(false)
+			if i == threshold-1 {
+				select {
+				case v := <-statusChange:
+					assert.Equal(t, AppStatusUnhealthy, v)
+				case <-time.After(100 * time.Millisecond):
+					t.Error("did not get a status change before deadline")
+				}
+				statusChange <- 255 // Fill the channel again
+			} else {
+				time.Sleep(time.Millisecond)
+			}
+		}
+	}
+
+	// Trigger threshold+10 failures and detect the invocation after the threshold
+	simulateFailures(threshold + 10)
+	assert.Empty(t, unexpectedStatusChanges.Load())
+	assert.Equal(t, threshold+10, h.failureCount.Load())
+
+	// First success should bring the app back to healthy
```

Comment:
@ItalyPaleAle Since the health probe waits for a threshold to be passed before considering the app is unhealthy, should there be a threshold also set to consider the app to be health after an unhealthy state? Can't there be cases where the app become healthy (health endpoint works) and immediately becomes unhealthy (probably restart of container, lazy load init issue etc)?

In these scenarios should we consider a threshold of successful health checks before we change status to healthy? 
cc @yaron2 @artursouza WDYT?

Reply:
A threshold ≥ 1 is common to detect failures, as it can happen that apps are just busy. Failure thresholds are particularly useful when there's a network involved as the network is unstable; that's not a big concern in this case as sidecar and app are in the same pod,  so perhaps a lower default threshold (2 or even 1) may even be an option.

A success threshold is a thing, but I don't believe it's a very common pattern. In the situations you describe, it sounds like the health endpoint would be implemented incorrectly, and the app should not report itself as healthy until all init work is performed.